dulwich/porcelain.py

```diff
@@ -191,6 +191,87 @@ class RemoteExists(Error):
     """Raised when the remote already exists."""
 
 
+def _parse_timezone_format(tz_str):
+    """Parse given string and attempt to return a timezone offset.
+    Different formats are considered in the following order:
+        - Git internal format: <unix timestamp> <timezone offset>
+        - RFC 2822: e.g. Mon, 20 Nov 1995 19:12:08 -0500
+        - ISO 8601: e.g. 1995-11-20T19:12:08-0500
+            - uses regex if dateutil is not available
+    Will default to UTC if timezone cannot be determined.
+    Args:
+      tz_str: datetime string
+    Returns: Timezone offset as integer
+    """
+    import re
+
+    # Git internal format
+    internal_format_pattern = re.compile("^[0-9]+ [+-][0-9]{,4}$")
+    if re.match(internal_format_pattern, tz_str):
+        try:
+            tz_internal = parse_timezone(tz_str.split(" ")[1].encode(DEFAULT_ENCODING))
+            return tz_internal[0]
+        except ValueError:
+            pass
+
+    # RFC 2822
+    import email.utils
+    rfc_2822 = email.utils.parsedate_tz(tz_str)
+    if rfc_2822:
+        return rfc_2822[9]
+
+    # ISO 8601
+    try:
+        from dateutil.parser import isoparse
+        try:
+            iso_8601 = isoparse(tz_str)
+            delta = iso_8601.tzinfo._offset
+            if delta:
+                # From dateutil doc:
+                # `The time zone offset in seconds, or (since version 2.6.0, represented as a datetime.timedelta object).`
+                if isinstance(delta, int):
+                    return delta
+                return int(delta.total_seconds())
+        except (AttributeError, TypeError, ValueError):
+            pass
+    except (ModuleNotFoundError, ImportError):
+        # Supported offsets:
+        # sHHMM, sHH:MM, sHH
+        iso_8601_pattern = re.compile("[0-9]([+-])([0-9]{2})(?::(?=[0-9]{2}))?([0-9]{2})?$")
+        match = re.search(iso_8601_pattern, tz_str)
+        total_secs = 0
+        if match:
+            sign, hours, minutes = match.groups()
+            total_secs += int(hours) * 3600
+            if minutes:
+                total_secs += int(minutes) * 60
+            total_secs = -total_secs if sign == "-" else total_secs
+            return total_secs
+
+    # YYYY.MM.DD, MM/DD/YYYY, DD.MM.YYYY contain no timezone information
+
+    return 0
```

Comment:
If the timezone is unparseable, then please throw an exception rather than return 0.

Reply:
Should we throw a custom Exception for this?

Reply:
Resolved in 33d9f65866cfffe212bec969f1a5afd79a8ebc7a